# 2단계. 데이터 수집 및 전처리

## 2. 데이터 불러오기

 컴퓨터에 저장된 데이터를 코랩에 업로드하기 위해 google.colab에서 files라는 패키지를 불러온 후, upload() 함수를 통해 myfile이라는 이름으로 저장한다.
 1. 'CO2 Emissions.csv' 파일을 불러와 코랩에 업로드하고 myfile에 저장한다.

In [1]:
#코랩 환경에 실습에 사용할 데이터 파일 불러오기
from google.colab import files
#코랩에 파일 불러와 myfile 변수에 저장하기
myfile = files.upload()

Saving CO2 Emissions.csv to CO2 Emissions.csv


코드를 실행하면 그림과 같이 '파일 선택' 버튼이 생성된다. 컴퓨터의 csv 파일을 선택해 업로드를 완료하면 '100% done'이라는 말과 함께 파일명이 뜬다.

2. 데이터 처리에 필요한 라이브러리인 pandas를 불러와 pd라는 별칭으로 줄여 사용한다. 그후 업로드한 데이터의 내용을 읽어 df로 저장하고, head()를 이용하여 내용을 확인한다.

In [2]:
#pandas 라이브러리를 불러와 pd라는 별칭 지정하기
import pandas as pd
#csv파일의 내용을 읽어와 df라는 변수에 저장하기
df = pd.read_csv("CO2 Emissions.csv")

3. df.head()를 이용해~~~ 현재 내용 그대로 작성

In [3]:
#head() 함수를 이용해 데이터셋의 상위 5줄 확인하기
df.head()

Make       Model Vehicle Class  Engine Size(L)  Cylinders Transmission  \
0  ACURA         ILX       COMPACT             2.0          4          AS5   
1  ACURA         ILX       COMPACT             2.4          4           M6   
2  ACURA  ILX HYBRID       COMPACT             1.5          4          AV7   
3  ACURA     MDX 4WD   SUV - SMALL             3.5          6          AS6   
4  ACURA     RDX AWD   SUV - SMALL             3.5          6          AS6   

  Fuel Type  Fuel Consumption City (L/100 km)  \
0         Z                               9.9   
1         Z                              11.2   
2         Z                               6.0   
3         Z                              12.7   
4         Z                              12.1   

   Fuel Consumption Hwy (L/100 km)  Fuel Consumption Comb (L/100 km)  \
0                              6.7                               8.5   
1                              7.7                               9.6   
2                              5.8                               5.9   
3                              9.1                              11.1   
4                              8.7                              10.6   

   Fuel Consumption Comb (mpg)  CO2 Emissions(g/km)  
0                           33                  196  
1                           29                  221  
2                           48                  136  
3                           25                  255  
4                           27                  244

csv 파일이 어떤 형태로 구성되어 있는지 살펴보니, 자동차 제조업체명, 모델명, 모델명, 연료량 등의 속성이 포함된 정형 데이터인 것을 볼 수 있다.

## 3. 데이터 확인 및 전처리

1.df.info()는 데이터 속성의 자료형과 자료 개수를 알려 주는 함수이다. 이를 통해 데이터의 정보를 확인할 수 있다.

In [4]:
#데이터 속성의 자료형과 속성별 데이터 개수 알아보기
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7385 entries, 0 to 7384
Data columns (total 12 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Make                              7385 non-null   object 
 1   Model                             7385 non-null   object 
 2   Vehicle Class                     7385 non-null   object 
 3   Engine Size(L)                    7385 non-null   float64
 4   Cylinders                         7385 non-null   int64  
 5   Transmission                      7385 non-null   object 
 6   Fuel Type                         7385 non-null   object 
 7   Fuel Consumption City (L/100 km)  7385 non-null   float64
 8   Fuel Consumption Hwy (L/100 km)   7385 non-null   float64
 9   Fuel Consumption Comb (L/100 km)  7385 non-null   float64
 10  Fuel Consumption Comb (mpg)       7385 non-null   int64  
 11  CO2 Emissions(g/km)               7385 non-null   int64  
dtypes: flo

info()의 결과를 보아 7385개의 데이터로 이루어진 것을 알 수 있으며, 각 속성은 모두 정수형, 실수형 수칫값 또는 문자형으로 이루어졌다는 것을 알 수 있다. Dtype에서, int64는 정수형 수칫값, float64는 실수형 수칫값, object는 그 외 문자형 등의 값을 의미한다. 또한 non-null이라는 것으로 보아 결측치가 없는 데이터라고 할 수 있다.

#  3단계. 기계학습 유형과 알고리즘 선정

#  4단계. 기계학습을 통한 모델 생성

학습 모델을 마들기 위해~~ 부터 다 그대로 쓰되, (1)의 코드를 아래 코드로 교체

In [15]:
#학습할 속성이 될 X(독립변수)와 타겟(목표)속성이 될 y(종속변수)를 나누어 저장하기
X = df[['Engine Size(L)', 'Fuel Consumption Comb (L/100 km)']]
y = df['CO2 Emissions(g/km)']

(2) 훈련 데이터와 ~~~~ 도 그대로 작성

In [16]:
#훈련 데이터와 테스트 데이터를 나누기
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=10)

(3) 도 그대로 작성, 하지만 실행결과가 작으니 세로로 나열하지 않고 코드 오른쪽에 가로로 (2단) 넣어도 공간 효율적으로 구성 가능할 듯 합니다.

In [17]:
#선형회귀를 이용해 모델 생성하기
from sklearn.linear_model import LinearRegression

linear_model = LinearRegression()  #선형회귀 모델 생성하기
linear_model.fit(X_train, y_train) #선형회귀 모델 학습하기

LinearRegression()

(4)

In [18]:
#학습 모델을 이용해 X_test 데이터에 대한 y 기대값을 예측한 결과를 linear_predict 변수에 저장하기
linear_predict = linear_model.predict(X_test)

In [19]:
#예측 결과가 들어있는 linear_predict 값 확인해보기
linear_predict

array([317.44207752, 279.08799323, 338.86253731, ..., 243.03871947,
       306.34999328, 254.23990494])

(5) 가상의 ~~

In [25]:
#가상의 독립 변수 값을 넣어 예측값이 어떻게 되는지 확인해보기
# 예) 엔진 크기가 3, 평균 연료소비량이 7.5일 경우의 이산화 탄소 배출량 예측해보기
linear_model.predict([[3.0,7.5]])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([200.71150049])

# 5단계. 성능 평가

사이킷런에는 기계학습 모델의 성능 평가를 하기 위한 함수가 포함되어 있다. R2 score을 계산하기 위한 함수인 r2_score 함수를 사용한다.

In [22]:
from sklearn.metrics import r2_score

In [23]:
r2_score(y_test, linear_predict)

0.8647792980764137

In [24]:
linear_model.score(X_test, y_test)

0.8647792980764137